### PR TITLE
Merge remote-tracking branch 'origin/master' into aha/FDP

### PR DESCRIPTION
**Pull request by andyHa**

### CODE REVIEW

File: src/main/java/sirius/biz/storage/VirtualObject.java [hunk 1]
```diff
@@ -192,7 +205,7 @@ public String getPhysicalKey() {
     public void setPhysicalKey(String physicalKey) {
         this.physicalKey = physicalKey;
     }
-
+@Override
```
Comment (Tobsucht) [inline src/main/java/sirius/biz/storage/VirtualObject.java]: reformate

File: src/main/java/sirius/biz/storage/vfs/SystemRoot.java [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+ * Provides some health and debugging data for sys admins.
+ */
+@Register
+public class SystemRoot implements VFSRoot {
+
+    @Part
+    private Metrics metrics;
+
+    @Override
+    public void collectRootFolders(VirtualFile parent, Consumer<VirtualFile> consumer) {
```
Comment (Tobsucht) [inline src/main/java/sirius/biz/storage/vfs/SystemRoot.java]: better name for consumer?

File: src/main/java/sirius/biz/storage/vfs/VFSRoot.java [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+ */
+public interface VFSRoot {
+
+    /**
+     * Collects all top-level files provided by this root.
+     *
+     * @param parent   the root directory of the VFS to be used a parent for the provided files
+     * @param consumer the consumer to collect all provided files
+     */
+    void collectRootFolders(VirtualFile parent, Consumer<VirtualFile> consumer);
```
Comment (Tobsucht) [inline src/main/java/sirius/biz/storage/vfs/VFSRoot.java]: parameter naming (consumer)

File: src/main/java/sirius/biz/storage/vfs/VirtualFile.java [hunk 4]
```diff
@@ -0,0 +1,334 @@
...
+import java.util.Optional;
+import java.util.function.BiConsumer;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+/**
+ * Represents a file or directory in the {@link VirtualFileSystem}.
+ * <p>
+ * The way this file can be modified, can be controlled by providing the appropriate callbacks. No no appropriate
```
Comment (Tobsucht) [inline src/main/java/sirius/biz/storage/vfs/VirtualFile.java]: No no -> If no

File: src/main/java/sirius/biz/storage/vfs/VirtualFile.java [hunk 5]
```diff
@@ -0,0 +1,334 @@
...
+
+        return deleteHandler.get();
+    }
+
+    /**
+     * Enumerates all children of this file.
+     *
+     * @param consumer the consumer collecting all children of this file
+     */
+    public void enumerateChildren(Consumer<VirtualFile> consumer) {
```
Comment (Tobsucht) [inline src/main/java/sirius/biz/storage/vfs/VirtualFile.java]: parameter naming

File: src/main/java/sirius/biz/storage/vfs/WorkRoot.java [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+                        storage.findOrCreateObjectByPath(UserContext.getCurrentUser().as(Tenant.class), WORK, name);
+                return storage.updateFile(newFile);
+            });
+        }
+
+        workDir.withChildren(this::listChildren);
+        return workDir;
+    }
+
+    private void listChildren(VirtualFile parent, Consumer<VirtualFile> consumer) {
```
Comment (Tobsucht) [inline src/main/java/sirius/biz/storage/vfs/WorkRoot.java]: parameter naming -> consumer

File: src/main/java/sirius/biz/storage/vfs/ftp/BridgeFileSystemView.java [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+
+    @Parts(VFSRoot.class)
+    private static Collection<VFSRoot> roots;
+
+    BridgeFileSystemView() {
+        root = VirtualFile.createRootNode().withChildren(this::computeRoots);
+        cwd = root;
+    }
+
+    private void computeRoots(VirtualFile parent, Consumer<VirtualFile> consumer) {
```
Comment (Tobsucht) [inline src/main/java/sirius/biz/storage/vfs/ftp/BridgeFileSystemView.java]: parameter naming -> consumer

File: src/main/java/sirius/biz/storage/vfs/VirtualFile.java [hunk 8]
```diff
@@ -0,0 +1,334 @@
...
+            return null;
+        }
+
+        return outputStreamSupplier.get();
+    }
+
+    /**
+     * Determines if the file is readable.
+     *
+     * @return <tt>true</tt> if the file is readable, false otherwise
```
Comment (fkrone) [inline src/main/java/sirius/biz/storage/vfs/VirtualFile.java]: tt is missing for false

File: src/main/java/sirius/biz/storage/vfs/VirtualFile.java [hunk 9]
```diff
@@ -0,0 +1,334 @@
...
+     */
+    public boolean isReadable() {
+        return inputStreamSupplier != null;
+    }
+
+    /**
+     * Determines if the file is writeable.
+     *
+     * @return <tt>true</tt> if the file is writeable, false otherwise
```
Comment (fkrone) [inline src/main/java/sirius/biz/storage/vfs/VirtualFile.java]: tt is missing for false

**Pull request merged**